### DISCUSSION TITLE
Dependency cycle in tag-driven CI/CD

### DISCUSSION THREAD

**Discussion by haraldatbmw** [fluxcd/flux2#500] [Q&A] [answered]
I just came up with a dependency cycle which I want to discuss here.

1. In my app-repo I store both my **app sources** and my **k8s manifests with kustomization overlays**.
2. A commit on my app-repo master branch triggers a build which ends with a new docker-image (tagged with build-number, e.g. 1.0.0-build25) in my docker-registry.
3. Flux **image-reflector-controller** and **image-automation-controller** detect the new image-tag in the registry and updates+pushes my kustomization.yaml with the new image-tag **which again trigges my build-pipeline... (2.)**

So I created an endless loop by accident.

To fix this I need to move all k8s manifests with kustomization overlays to my git-ops repo where no Jenkins build-pipeline is triggered. So this should work.

Are there any other options to bypass such a situation?

**Comment by stefanprodan** [comment, marked as answer]
Most CI tools will ignore commits with `[skip ci]` as the prefix of commit message, if Jenkins doesn't do that, I think it's easy to add it. 

I think you can add `[skip ci]` to the commit message inside the image automation resource cc @squaremo

**Reply by haraldatbmw** [reply]
Okay this would overcome the endless loop. 

But one messy thing remains. The commit-id which created the docker-image (due to the jenkins build) is not the same that flux uses for deployment. Deployment manifests are one commit ahead (due to image-automation-controller).

My integration-environment has a git-repository source which uses a pre-release semver range for the app-repo.
So the image-automation-controller should tag the repo after push which feels ugly too.

So I prefer putting my application manifests to the git-ops repo and work in this repo without tags.
And using semver ranges on the image-reflector-controller instead.

**Reply by squaremo** [reply]
What Stefan suggested is what Flux v1 does; its commit messages include `[skip ci]` (or a similar token you nominate). I could include it in the default message for image automation commits -- that would probably avoid 90% of these loops, in one stroke.

(logged as https://github.com/fluxcd/image-automation-controller/issues/52)

**Reply by haraldatbmw** [reply]
Yes, this commit message would be fine to prevent loops.

But my use-case of using pre-release and release tags for triggering deployment to test or prod environment are not solved by that.
I tag a commit in the app-repo to force my Jenkins-pipeline to retag an existing docker-image with a semantic version tag (pre-release). This new image-tag is detected by the image-reflector and updated by by image-automation in the same repo. To get this new manifest change detected by my flux git source it can not work with a semver range because the commit of image-automation is not tagged. So the git source fetches every commit and I have lost my environment gateway using git-tags. So any other change, e.g. inside a kustomize-base yaml is deployed to all environments on commit. 

Can you follow my explanation?

EDIT: I added a diagram for better understanding.

![cicd-flow](https://user-images.githubusercontent.com/49680586/100141671-11490380-2e93-11eb-96bd-e1e716050c5e.png)

**Reply by haraldatbmw** [reply]
For my use-case it would make sense that the image-automation-controller can be configured to create a git-tag on the commit.

**Reply by stefanprodan** [reply]
Why not create an overlay per environment and configure the image policy to update the dev overlay for pre-releases and the prod overlay with releases? example here: https://github.com/fluxcd/flux2-kustomize-helm-example/tree/main/apps

**Reply by haraldatbmw** [reply]
Yes for sure overlays on different environments. 
But if I commit a change to one base yaml which is used in every overlay. This commit would go to dev, test & production at commit-time. So I have to prevent this by adding a semver gateway on the git-source for test (pre-releases) and production (releases). With that in place I need an additional manual tag on the Ops-Repo too.

**Reply by squaremo** [reply]
I think I see (the diagram helps!) ... You are deploying semver-tagged versions of the git repo, and you want an automated update to bump the semver tag, so the change it made gets deployed.

Could you have CI notice the push from the image automation, and tag it?

**Reply by phillebaba** [reply]
I think having CI do these tasks is a much simpler solution than having flux implement this logic. I have done something similar to this by triggering CI pipelines that commit and tag to different repositories.

**Reply by haraldatbmw** [reply]
Yes a CI pipeline for the Ops-Repo could be triggered by the push from the image-automation-controller. But for auto-tagging it this pipeline has to decide if the commit was from flux image-automation and extract the tag-name out of anywhere...

**Reply by haraldatbmw** [reply]
Just to make it clear the CI logic has to do this:

**pseudo-code:**
```
...
var commitUser = extractCommitUser();

if (commitUser == 'flux-image-automation') {
  var tagName = extractTagNameFromCommit();

  if (isPreReleaseTag(tagName) || isRelease(tagName)) {
    createAndPushTag(tagName);
  }
}
...
```

**Reply by squaremo** [reply]
I missed the bit where the automation decides what the tag name is -- does it get named for the image that's been updated, or bumped from the last such tag, or ..?

**Reply by haraldatbmw** [reply]
In my case the app-repo git-tag is also the image-tag which can be used to create the tag in the ops-repo.
So the automation knows the image-tag?

**Reply by squaremo** [reply]
This feels like quite a specific use case*, and I'm reluctant to add specialised features to a controller. But I do want to get to a point where you have a plausible way forward @haraldatbmw. (*The specific bit being: the deployments are driven by a git tag which is exactly the image tag of a particular image.)

If we could make it possible for you to set up a CI action something like the pseudo-code you posted above, would that be good enough do you think? I am thinking mainly about being able to reliably get the image tag in question from the commit message; e.g., by providing template values for each image that was updated in the commit, so you can include them.

**Reply by haraldatbmw** [reply]
@squaremo I think the CI action should be fine for me. It would be nice if the commit-message of the image-automation-commit would contain the value of the replaced image-policy in the yaml file. So my CI action can parse this info and reuse it for tagging.
